src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/HttpGoogleCloudStorageDAO.scala

```diff
@@ -1,81 +1,245 @@
+import scala.collection.JavaConversions._
+import scala.collection.JavaConverters._
+import scala.util.{Failure, Success, Try}
+
+// Seq[String] -> Collection<String>
+
+class HttpGoogleCloudStorageDAO(
+  useServiceAccountForBuckets: Boolean,
+  clientSecretsJson: String,
+  dataStoreRoot: String,
+  rawlsRedirectBaseUrl: String,
+  p12File: String,
+  appsDomain: String,
+  groupsPrefix: String,
+  appName: String) extends GoogleCloudStorageDAO {
+
+  val groupMemberRole = "MEMBER" // the Google Group role corresponding to a member (note that this is distinct from the GCS roles defined in WorkspaceAccessLevel)
+
+  val groupAccessLevelsAscending = Seq(WorkspaceAccessLevel.Read, WorkspaceAccessLevel.Write, WorkspaceAccessLevel.Owner)
 
-class HttpGoogleCloudStorageDAO(clientSecretsJson: String, dataStoreFactory: DataStoreFactory, ourBaseURL: String)(implicit system: ActorSystem) extends GoogleCloudStorageDAO {
   // modify these if we need more granular access in the future
-  val gcsFullControl = "https://www.googleapis.com/auth/devstorage.full_control"
-  val computeFullControl = "https://www.googleapis.com/auth/compute"
-  val scopes = Seq(gcsFullControl,computeFullControl)
+  val storageScopes = Seq(StorageScopes.DEVSTORAGE_FULL_CONTROL, ComputeScopes.COMPUTE)
+  val directoryScopes = Seq(DirectoryScopes.ADMIN_DIRECTORY_GROUP)
 
   val httpTransport = GoogleNetHttpTransport.newTrustedTransport
   val jsonFactory = JacksonFactory.getDefaultInstance
   val clientSecrets = GoogleClientSecrets.load(jsonFactory, new StringReader(clientSecretsJson))
+  val dataStoreFactory = new FileDataStoreFactory(new File(dataStoreRoot))
 
-  val flow = new GoogleAuthorizationCodeFlow.Builder(httpTransport, jsonFactory, clientSecrets, scopes).setDataStoreFactory(dataStoreFactory).build()
+  val flow = new GoogleAuthorizationCodeFlow.Builder(httpTransport, jsonFactory, clientSecrets, storageScopes).setDataStoreFactory(dataStoreFactory).build()
 
-  override def getOurRedirectURI(callbackPath: String): String = s"${ourBaseURL}/${callbackPath}"
+  override def getRawlsRedirectURI(callbackPath: String): String = s"${rawlsRedirectBaseUrl}/${callbackPath}"
 
   override def getGoogleRedirectURI(userId: String, callbackPath: String): String = {
     flow.newAuthorizationUrl()
-     .setRedirectUri(getOurRedirectURI(callbackPath))
+     .setRedirectUri(getRawlsRedirectURI(callbackPath))
      .setState(userId)
      .setAccessType("offline")   // enables token refresh
      .build()
   }
 
+  override def setupACL(ownerId: String, bucketName: String, workspaceName: WorkspaceName): Unit = {
+    val bucketCredential = getBucketCredential(ownerId)
+    // ACLs are stored internally in three Google groups.
+    groupAccessLevelsAscending foreach { level => createGoogleGroupWithACL(bucketCredential, level, bucketName, workspaceName) }
+
+    // add the owner -- may take up to 1 minute after groups are created
```

Comment:
see https://broadinstitute.atlassian.net/browse/DSDEEPB-1114. I haven't researched it, based on my testing there's no delay in the call finishing, and the person shows up immediately in the groups UI.